src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+
+use crate::wal::{encode_progress, encode_update};
+
+/// Organize and maintain a compacted representation of persisted data.
+///
+/// The `Compacter` task keeps track of the persisted updates for each persisted relation and
+/// periodically compacts that representation to use space proportional to the number of
+/// distinct rows in the relation. In order to do so, the `Compacter` maintains a `Trace` for
+/// each persisted relation.
+///
+/// A `Trace` is basically a list of `Batch`s that represent a contiguous time interval, and a
+/// compaction frontier.
+///
+/// A `Batch` is a consolidated list of updates that occured between times [lower, upper)
+/// where each update is of the form `(Row, time, diff)` and each `(Row, time)` pair occurs
+/// exactly once and all diffs are nonzero.
+///
+/// Note that all `Batch`s keep their data on persistent storage. No data resides on disk
+/// (except currently we load all the data from batches into memory for compaction and on
+///  restart but this will get fixed!).
+///
+/// The `Coordinator` thread tells the `Compacter` when it needs to
+///  * start keeping track of a new relation
+///  * stop keeping track of a relation
+///  * resume keeping track of a relation with an already initialized Trace (on restart)
+///  * advance a relation's compaction frontier. Note that this doesn't automatically trigger
+///    any actual compaction. That happens later (keep reading).
+///
+/// The `Compacter` task periodically checks each relation's WAL directory to look
+/// for finished log segments, converts them to `Batch`s (basically consolidates the
+/// updates for a range of times) and adds them to the relation's `Trace`.
+///
+/// When a `Trace` contains too many `Batches`, the Trace physically
+/// combines all of them into a single large batch with updates
+/// forwarded to the compaction frontier.
+
+// TODO: Lets add some jitter to compaction so we aren't compacting every single
+// relation at the same time maybe?
+static COMPACTER_INTERVAL: Duration = Duration::from_secs(300);
+
+/// Instructions that the Coordinator sends the Compacter.
+#[derive(Debug)]
+pub enum CompacterMessage {
+    Add(GlobalId),
+    Drop(GlobalId),
+    Resume(GlobalId, Trace),
+    AllowCompaction(GlobalId, Timestamp),
+}
+
+/// A Batch contains all of the updates that originated within some time range [lower, upper)
+/// but the data live the file at `path`.
+///
+/// The data stored in each batch are triples of (Row, Timestamp, Diff) such that there is
+/// exactly one copy of each (Row, Timestamp) in each batch. Batches also have a header
+/// and a footer indicating the upper and lower bound timestamps.
+/// TODO: Batches are meant to mimic differential / cdcv2 batches but do not do so
+/// currently. Let's fix that. Specifically, introduce a `since` field, and counts
+/// for the number of updates at each time.
+#[derive(Debug)]
+struct Batch {
+    upper: Timestamp,
+    lower: Timestamp,
```

Comment:
If it matters / helps, there is a specific DD struct `Description` for this, that handles some corner cases. Most obviously multi-dimensional times, but also an empty `upper` which indicates completeness (e.g. with a file input, or some other way folks have to supply data). I could imagine "this is just for `TABLE`" but the more different places we have to home the same logic the worse, I think. Also, imagine someone wanted to add a `CLOSE TABLE` or whatever method, to improve perf in joins and such; it would suck if this blocked that.

Reply:
added a todo